### PR TITLE
WIP: Corpus class

### PR DESCRIPTION
**Pull request by eseiver**
This PR is not ready to merge yet.

It includes a proposal for how a Corpus class might work and some initial methods for a class.

### CODE REVIEW

**Comment by eseiver** [comment]
@mpacer would love to get your thoughts on docs & class

**Comment by eseiver** [comment]
Thanks for checking Sebastian! Still a work in progress though, no need to merge.



~Elizabeth
PLOS researcher on-the-go
415-902-6389


-------- Original message --------

**Comment by sbassi** [comment]
ok, so please put the nomerge label (if not there)

On Dec 30, 2017 12:30 AM, "Elizabeth Seiver" <notifications@github.com>
wrote:

Thanks for checking Sebastian! Still a work in progress though, no need to
merge.



~Elizabeth
PLOS researcher on-the-go
415-902-6389 <(415)%20902-6389>


-------- Original message --------

File: allofplos/article_class.py [hunk 1]
```diff
@@ -223,6 +223,7 @@ def get_plos_journal(self, caps_fixed=True):
                                                                 "journal-title"])
             journal = journal[0].text
         except IndexError:
+            # Need to file JIRA ticket: only affects pone.0047704
```
Comment (mpacer) [inline allofplos/article_class.py]: how is this related to the corpus class?

File: allofplos/corpus_class.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        self.plos_network = plos_network
+        self.extension = extension
+
+    @property
+    def directory(self):
+        """
+        :returns: directory of corpus
+        :rtype: {str}
+        """
+        return self._directory
```
Comment (mpacer) [inline allofplos/corpus_class.py]: This should check that it actually is a directory first.

File: allofplos/corpus_class.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+    @directory.setter
+    def directory(self, d):
+        """
+        Reset memoized info when changing the path to the corpus.
+        """
+        self._directory = d
```
Comment (mpacer) [inline allofplos/corpus_class.py]: This should check that it actually is a directory first. 

If at any point you specify a spec for these corpus directories, it should also check to make sure that spec is followed.

File: allofplos/corpus_class.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+    def __repr__(self):
+        """Value of a corpus object when you call it directly on the command line.
+
+        Shows the directory location of the corpus
+        :returns: directory
+        :rtype: {str}
+        """
+        out = "Corpus location: {0}\nNumber of files: {1}".format(self.directory, len(self.files))
```
Comment (mpacer) [inline allofplos/corpus_class.py]: cute!

**Comment by mpacer** [comment]
I'm going to make a bunch of comments about what should be removed for the initial corpus class merge. I think it'll help to have the most basic features that will help people today, and then to iterate on top of that to add more advanced things (e.g., a hash table management system).

File: allofplos/corpus_class.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    @property
+    def dois(self):
+        """List of DOIs of the articles in the corpus directory."""
+
+        return list(self.iter_dois)
+
+    @property
+    def filepaths(self):
+        """List of article XML files in corpus directory, including the full path."""
+        return [os.path.join(self.directory, fname) for fname in self.files]
```
Comment (mpacer) [inline allofplos/corpus_class.py]: is this supposed to be the abs path or the relative path?

File: allofplos/corpus_class.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        :param directory: directory of articles for class object
+        :param count: specify how many DOIs are to be returned
+        :return: a list of random DOIs for analysis
+        """
+        doi_list = self.dois
+        sample_doi_list = random.sample(doi_list, count)
+
+        return sample_doi_list
+
+    def hashtable(self, directory=directory):
```
Comment (mpacer) [inline allofplos/corpus_class.py]: I think this functionality is advanced and should be factored out into its own PR.

File: allofplos/corpus_class.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        return out
+
+    @property
+    def iter_file_doi(self):
+        """Generator that returns filename, doi tuples for every file in the corpus.
+
+        Used to generate both DOI and file generators for the corpus. 
+        """
+        return ((file_, filename_to_doi(file_))
+                for file_ in os.listdir(self.directory)
```
Comment (mpacer) [inline allofplos/corpus_class.py]: If you have any other files in the directory that do not match the filename_to_doi expected pattern, this may behave strangely. I'd appreciate if you could make some of those checks more explicit (which unfortunately means probably breaking up that pretty generator comprehension into a couple of steps).

File: allofplos/corpus_class.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+        :return: dict of article DOIs to hashes
+        """
+        hash_dict = {}
+        for fname in tqdm(self.filepaths):
+            file_hash = hash_file(fname)
+            hash_dict[filename_to_doi(fname)] = file_hash
+
+        return OrderedDict(hash_dict)
+
+    def hashtable_to_json(self, hashtable=None, overwrite=True):
```
Comment (mpacer) [inline allofplos/corpus_class.py]: Refactor to hashtable pr.

File: allofplos/corpus_class.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+        if hashtable is None:
+            hashtable = self.hashtable()
+        if overwrite is True or os.path.isfile(hash_json) is False:
+            with open(hash_json, 'w') as fp:
+                json.dump(hashtable, fp)
+        else:
+            print("aborting: .json hashtable exists. To overwrite, set parameter 'overwrite' to False.")
+        return hashtable
+
+    def read_hashtable(self):
```
Comment (mpacer) [inline allofplos/corpus_class.py]: refactor to hashtable pr

File: allofplos/corpus_class.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+        Also see `self.hashtable{}`
+        """
+        if os.path.isfile(hash_json):
+            return OrderedDict(json.load(open(hash_json)))
+        else:
+            print("Hashtable not found; create with Corpus().hashtable()")
+            return None
+
+    def update_hashtable(self, update_json=True):
```
Comment (mpacer) [inline allofplos/corpus_class.py]: refactor to hashtable pr

File: allofplos/corpus_class.py [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+        if update_json:
+            json_dump = self.hashtable_to_json(hashtable=resorted_hashtable)
+
+        else:
+            print('Hashtable .json file not found. Create with `Corpus().hashtable_to_json()`')
+            resorted_hashtable = {}
+
+        return resorted_hashtable
+
+    # @symlinks.setter
```
Comment (mpacer) [inline allofplos/corpus_class.py]: what is this doing here? Do you still want it to exist?

File: allofplos/corpus_class.py [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+    @classmethod
+    def create(cls, directory=corpusdir, overwrite=False):
+        """ Create a PLOS corpus from scratch.
+
+        Downloads and unzips the main PLOS .zip file hosted on Google Drive.
+        :param directory: directory to download and unzip zip file
+        :param overwrite: whether to overwrite existing files, defaults to False
+        """
+        os.makedirs(directory, exist_ok=True)
+        if overwrite is False and len(os.listdir(directory)) > 2:
```
Comment (mpacer) [inline allofplos/corpus_class.py]: why are you checking that it has more than two files? what's special about 2?

File: allofplos/corpus_class.py [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+    #     """Sets a corpus object using a set of dois.
+
+    #     Converts a filename to DOI using an existing function.
+    #     :param value: filename
+    #     :type value: string
+    #     """
+    #     self.doi = filename_to_doi(value)
+
+    @classmethod
+    def create(cls, directory=corpusdir, overwrite=False):
```
Comment (mpacer) [inline allofplos/corpus_class.py]: I'm not 100% sure that this should be part of the initial corpus PR… it seems like this is a really useful but potentially complciated feature… it could be part of the first PR that is merged about the corpus after the corpus itself has been a bit more battle tested.

File: allofplos/corpus_class.py [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+            zip_date, zip_size, metadata_path = get_zip_metadata()
+            zip_path = download_file_from_google_drive(zip_id, local_zip, file_size=zip_size,
+                                                       destination=directory)
+            unzip_articles(file_path=zip_path, extract_directory=directory)
+            os.remove(metadata_path)
+            print("Corpus created with {} files".format(len(os.listdir(directory))))
+            return cls(directory=directory)
+
+    @classmethod
+    def from_dois(cls, dois, source=corpusdir, destination='testdir', overwrite=True):
```
Comment (mpacer) [inline allofplos/corpus_class.py]: This definitely should be in its own PR.

File: allofplos/corpus_helpers.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+import requests
+
+from .transformations import BASE_URL_API
+
+BLOCKSIZE = 65536
+hasher = hashlib.sha256()
+
+
+def hash_file(fname):
```
Comment (mpacer) [inline allofplos/corpus_helpers.py]: move to hashtable pr

File: allofplos/corpus_helpers.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        buf = afile.read(BLOCKSIZE)
+        while len(buf) > 0:
+            hasher.update(buf)
+            buf = afile.read(BLOCKSIZE)
+        file_hash = hasher.hexdigest()
+
+    return file_hash
+
+
+def search_solr_records(days_ago=14, start=0, rows=1000, start_date=None, end_date=None, item='id'):
```
Comment (mpacer) [inline allofplos/corpus_helpers.py]: This isn't specific to the corpus but seems to be an interactive api (not a library api); it probably should exist not only in a separate PR but possibly a separate module.

File: allofplos/corpus_helpers.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+    if solr_search_results:
+        print("{0} results returned from this search."
+              .format(len(solr_search_results)))
+    else:
+        print('No results returned for this search.')
+    return solr_search_results
+
+
+def get_all_solr_dois():
```
Comment (mpacer) [inline allofplos/corpus_helpers.py]: This should be in the same utility PR as `search_solr_records`. This seems wholly unrelated to the Corpus class.

File: allofplos/gdrive.py [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+local_zip = 'allofplos_xml.zip'
+zip_metadata = 'zip_info.txt'
+time_formatting = "%Y_%b_%d_%Hh%Mm%Ss"
+min_files_for_valid_corpus = 200000
+test_zip_id = '12VomS72LdTI3aYn4cphYAShv13turbX3'
+local_test_zip = 'sample_corpus.zip'
+gdrive_url = "https://docs.google.com/uc?export=download"
+
+
+def download_file_from_google_drive(id, filename, destination=corpusdir,
```
Comment (mpacer) [inline allofplos/gdrive.py]: This entire file should be moved into its own PR as it is completely unrelated to the Corpus class. This should be merged before the corpus class though, since the Corpus class depends on it.
